debug / workarounds for image extended tests wrt current ICT based de…

gabemontero opened the pull request:
…ploy behavior

gabemontero commented:
[testonlyextended][extended:core(images)]

gabemontero commented:
[testonlyextended][extended:core(images)]

gabemontero commented:
@bparees an FYI wrt images extended tests:

1) with this timing window wrt Deployments with ICTs and the associated container-> images field, over the last few days, the current behavior seen is that if container->images wins the race (which is not always the case ... sometimes the ICT wins the race), then we go through the error pulling image cycle, where the first deployment fails, and then a new deployment version is automatically started, the ICT supplies the image ref, and the deployment succeeds

the wait for a deployment logic in the extended tests previously could not cope with this.

This PR currently has changes which enable a retry in that logic, giving the second deployment a chance to succeed.  Test runs in this PR have confirmed the retry is working.

2) I'm seeing intermittent issues with even *starting* builds ... not sure how to debug that yet

3) I'm seeing intermittent issues verifying the service is up after the deployment ... this probably will be my next area of focus.

bparees commented:
> This PR currently has changes which enable a retry in that logic, giving the second deployment a chance to succeed. Test runs in this PR have confirmed the retry is working.

guessing we'll ultimately want to revert that change, but only after @kargakis' changes go in that make the " " image field values work "correctly".

> I'm seeing intermittent issues with even starting builds ... not sure how to debug that yet

what's the output from oc start-build, assuming that is where it fails?  or does the build pod get created, but never actually run?

gabemontero commented:
On Wed, May 18, 2016 at 2:19 PM, Ben Parees <notifications@github.com>
wrote:

> This PR currently has changes which enable a retry in that logic, giving
> the second deployment a chance to succeed. Test runs in this PR have
> confirmed the retry is working.
>
> guessing we'll ultimately want to revert that change, but only after
> @kargakis <https://github.com/kargakis>' changes go in that make the " "
> image field values work "correctly".
>
That was my assumption/hope as well.  But if I had to guess, there will
definitely be an interim period where we'll want to go with this change to
attack the next layer of problems.

> I'm seeing intermittent issues with even starting builds ... not sure how
> to debug that yet
>
> what's the output from oc start-build, assuming that is where it fails? or
> does the build pod get created, but never actually run?
>
Don't know yet precisely.  To date, all I know is that the `oc logs -f
bc/<bc name>` after the build failure fails as well.  Making sure the
start-build output is dumped
and running some `oc get pods` to get a lay of the land were my next
thoughts.

> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/8911#issuecomment-220113909>
>

kargakis commented:
> guessing we'll ultimately want to revert that change, but only after @kargakis' changes go in that make the " " image field values work "correctly".

Not sure why " " is special, my pull will fix all initial deployments with ICTs independetly of what's the initial image name in the podtemplate.

kargakis commented:
> Not sure why " " is special, my pull will fix all initial deployments with ICTs independetly of what's the initial image name in the podtemplate.

And independetly of automatic

gabemontero commented:
" " (with a single space) was just a work around.  @smarterclayton had asked us to change all the example / sample templates with deployment ICTs so that the container image field was "" (no space) vs. the dummy strings we used to have to better illustrate that the template wants the image to come from the ICT , but that fails k8s validation (which requires a value with len > 0 for that string)

gabemontero commented:
@bparees on the failed build with no build logs in the images extended tests, the build is suppose to be generated as the result of a `oc new-app`, which returned without error.

There are some spots in the extended test wait for build logic that could benefit from not eating errors.  I'll start there.

gabemontero commented:
@bparees @csrwng an update on the images extended tests where `oc new-app` intermittently does not start a build.  With the latest round of debug, I temporarily try an `oc start-build` when a build fails, and it fails as well, but with a more informative (but very perplexing) message:

`May 19 18:36:26.087: INFO: Running 'oc start-build --namespace=extended-test-s2i-perl-i04pz-7ebtd --config=/tmp/extended-test-s2i-perl-i04pz-7ebtd-user.kubeconfig dancer-mysql-example'
May 19 18:36:26.185: INFO: Error running &{/data/src/github.com/openshift/origin/_output/local/bin/linux/amd64/oc [oc start-build --namespace=extended-test-s2i-perl-i04pz-7ebtd --config=/tmp/extended-test-s2i-perl-i04pz-7ebtd-user.kubeconfig dancer-mysql-example] []   Error from server: imagestreams "perl" not found
 Error from server: imagestreams "perl" not found
`

Rest assured the "perl" imagestream was created at the start of the extended test run, and I saw no indication of it getting deleted.

Ever see something like this before?  Thoughts?

gabemontero commented:
@bparees @csrwng on the images extended tests where it looked like `oc new-app` was not starting deploys, with some additional debug last night, it looks like it is just another timing permutation with the image race condition wrt ICTs.

So unlike the build / missing imagestream mystery in my last comment, I'm punt on further work and simply retry with the recently merged improvements here that avoid the race condition.

bparees commented:
@gabemontero I wonder if the perl imagestream exists, but perhaps has not imported any tags (in which case the error message is misleading).  Probably need to track down where in the code we throw that error and see what conditions could lead to it (ie whether a lack of tags could do it)

it's also not clear which namespaces it's looking for the "perl" imagestream in?  i'm going to assume it's looking in the "current" namespace as opposed to the "openshift" one, since it doesn't indicate.  Not sure which namespace you were looking at?

Also the deployment/ICT race condition should be fixed now based on the PR that just merged today, so maybe rerun the tests and see, before adding retry logic.

gabemontero commented:
RE: which namespace the languange IS's are in ... I thought they still went in the openshift namespace (and I believe the other templates are looking there) ... I'm adding debug which will dump the ISs in both the "openshift" namespace and the dynamically allocated namespace each extended test gets.

The "perhaps has not imported any tags" theory seems plausible, I'll look into it.

And yes, the "recently merged improvements" in my last comment were the fixes to the deployment/ICT race condition.  I've already to a test cooking with those (while I sort out the aforementioned next set of debug).

gabemontero commented:
@bparees I'd like to go ahead and merge this PR.  It should get us cleaner runs on the jenkins job.

I'll mention up front it does have the workaround of seeing if the language imagestreams in the openshift namespace have been deleted, and if need be, recreates them.  My intent here is to get hopefully get cleaner runs on the jenkins job, further the path coverage, and with the debug, see if we can pin down when the imagestreams get deleted from job to job - narrow the search window.  Over time, if it becomes less prevalent, we can either remove, or change the code to simply fail the test if they are not there.

bparees commented on test/extended/util/framework.go:
```diff
@@ -318,12 +398,27 @@ func WaitForADeployment(client kclient.ReplicationControllerInterface, name stri
-			err, matched := okOrFailed(&rc)
-			if matched {
-				return err, false
+			if lastRC <= rc.GetName() {
```
i'm guessing we're not too worried about it, but this will fail if we get beyond 9 deployments.

gabemontero replied:
Yep, my thought was this was good enough, KISS.

bparees commented on test/extended/util/framework.go:
```diff
@@ -318,12 +398,27 @@ func WaitForADeployment(client kclient.ReplicationControllerInterface, name stri
 		for _, rc := range list.Items {
-			err, matched := okOrFailed(&rc)
-			if matched {
-				return err, false
```
i'm unclear on what happened here.  previously this code looped through all the deployments, breaking/returning as soon as it found one that was either complete or failed (ie okOrFailed returned matched==true).  That certainly seems wrong, but rather than checking okOrFailed on every deployment, would it make more sense for the new code to just find the most recent deployment and call okOrFailed on that one?

gabemontero replied:
I had it that way for a while, especially during the time frame when we were constantly getting bit by the ICT race condition.  Once that was resolved, I reverted to this simpler change (though in hindsight now I'm not sure why I was worried about introducing a bigger change).

So sure, I can add the "find the latest RC" element to this.

liggitt commented on test/extended/util/framework.go:
```diff
@@ -40,13 +40,79 @@ var TestContext e2e.TestContextType
+//FixOpenShiftNamespaceImageStreams is a temporary workaround for the intermittent
+// issue seen in extended tests where *something* is deleteing the pre-loaded, languange
+// imagestreams from the OpenShift namespace
+func FixOpenShiftNamespaceImageStreams(oc *CLI) {
```
Can we not add logging to the imagestream deletion rest handler or enable auditing to catch this, instead of this workaround?

gabemontero replied:
I was very hesitant to add logging in the mainline rest path, and wanted to do save that as a last resort.  And to reiterate, this workaround is going to be short term (I want to confirm that the better results I'm seeing in my pull with it translate to the jenkins job).

That said, enabling audit logging sounds promising to me though.  However, I'm having trouble finding how to do that in the documentation - any pointers?

thanks

gabemontero replied:
@liggitt and I agreed offline to convert the recreate imagestreams into the fail the test immediately idea I mentioned previously.  Also got a clue that the on/off switch for audit logging is in the master cfg.

gabemontero commented:
@bparees @liggit changes based on last set of comments pushed:
1) backed off recreating openshift namespace imagestreams, but will do a gomega/ginkgo error exit if we see they are missing
2) updated extended test setup script to enable audit logging on the master
3) updated wait for deployment to only call okOrFailed(rc) on the lastest rc (assuming we don't go past 9 versions)

bparees commented on test/extended/setup.sh:
```diff
@@ -94,6 +94,11 @@ function os::test::extended::setup {
     configure_os_server
+    #turn on audit logging for extended tests ... mimic what is done in util.sh configure_os_server, but don't
+    # put change there - only want this for extended tests
+    echo TURNING ON AUDIT LOGGING
```
echo "[INFO] Turning on audit logging"
(it's nit picky, but better to align w/ the existing output)

gabemontero replied:
update pushed

bparees commented on test/extended/util/framework.go:
```diff
@@ -318,12 +390,21 @@ func WaitForADeployment(client kclient.ReplicationControllerInterface, name stri
 		}
+
+		fmt.Fprintf(g.GinkgoWriter, "\n GGM checking status of rc %s \n", lastRC.GetName())
+		err, matched := okOrFailed(&lastRC)
```
lastRC could be uninitialized here since there's no guarantee that list.Items was non-empty.

bparees commented on test/extended/util/framework.go:
```diff
@@ -318,12 +390,21 @@ func WaitForADeployment(client kclient.ReplicationControllerInterface, name stri
-			if matched {
-				return err, false
+			// assuming won't have to deal with more than 9 deployments
+			if lastRC.GetName() <= rc.GetName() {
```
what gets returned by getName() when lastRC is uninitialized (first time through the loop)?  i guess it returns "" and that is less than everything?

gabemontero replied:
yep here too

gabemontero commented:
new updates pushed

bparees commented:
lgtm, would like to see at least a couple extended tests pass before merging so we know the new WaitForADeployment logic is working.

gabemontero commented:
the entire suite won't pass because of the known issues, but yes, let's see how a PR specific extended test run goes before putting in the merge comment

gabemontero commented:
[test][extended:core(images)]

gabemontero commented:
fyi @bparees i pushed the tweaks to get the postgresql-slave logs for https://github.com/openshift/origin/issues/9188 in the jenkins job, confirm they match my local repros.

bparees commented:
saw a couple passes in https://ci.openshift.redhat.com/jenkins/job/test_pr_origin_extended/212/console for tests using the new deploy logic.
[merge]
